src/parse.rs

```diff
@@ -421,82 +429,93 @@ impl<'a> Bytes<'a> {
-                b'a' | b'A' => n * 16_u16 + 10_u16,
-                b'b' | b'B' => n * 16_u16 + 11_u16,
-                b'c' | b'C' => n * 16_u16 + 12_u16,
-                b'd' | b'D' => n * 16_u16 + 13_u16,
-                b'e' | b'E' => n * 16_u16 + 14_u16,
-                b'f' | b'F' => n * 16_u16 + 15_u16,
-                _ => {
-                    return self.err(ParseError::InvalidEscape);
-                }
-            };
+            n = n << 4;
+            let byte = self.eat_byte()?;
+            n |= self.decode_hex(byte)? as u16;
         }
 
         Ok(n)
     }
+    */
+
+    fn decode_hex(&self, c: u8) -> Result<u8> {
+        match c {
+            c @ b'0'...b'9' => Ok(c - b'0'),
+            c @ b'a'...b'f' => Ok(10 + c - b'a'),
+            c @ b'A'...b'F' => Ok(10 + c - b'A'),
+            _ => self.err(ParseError::InvalidEscape("Non-hex digit found")),
+        }
+    }
 
     fn parse_str_escape(&mut self, store: &mut Vec<u8>) -> Result<()> {
         use std::iter::repeat;
 
         match self.eat_byte()? {
+            b'\'' => store.push(b'\''),
             b'"' => store.push(b'"'),
             b'\\' => store.push(b'\\'),
             b'b' => store.push(b'\x08'),
             b'f' => store.push(b'\x0c'),
             b'n' => store.push(b'\n'),
             b'r' => store.push(b'\r'),
             b't' => store.push(b'\t'),
+            b'x' => store.push(self.decode_ascii_escape()?),
             b'u' => {
-                let c: char = match self.decode_hex_escape()? {
-                    0xDC00...0xDFFF => {
-                        return self.err(ParseError::InvalidEscape);
-                    }
+                self.expect_byte(b'{', ParseError::InvalidEscape("Missing {"))?;
+
+                let mut bytes: u32 = 0;
 
-                    n1 @ 0xD800...0xDBFF => {
-                        if self.eat_byte()? != b'\\' {
-                            return self.err(ParseError::InvalidEscape);
-                        }
+                for _ in 0..6 {
+                    let byte = self.peek_or_eof()?;
 
-                        if self.eat_byte()? != b'u' {
-                            return self.err(ParseError::InvalidEscape);
-                        }
+                    if byte == b'}' {
+                        break;
+                    } else {
+                        self.advance_single()?;
+                    }
 
-                        let n2 = self.decode_hex_escape()?;
+                    let byte = self.decode_hex(byte)?;
+                    bytes = bytes << 4;
+                    bytes |= byte as u32;
+                }
 
-                        if n2 < 0xDC00 || n2 > 0xDFFF {
-                            return self.err(ParseError::InvalidEscape);
-                        }
+                if bytes == 0 {
+                    return self.err(ParseError::InvalidEscape(
+                        "Null character (\\u{0}) not allowed",
```

Comment:
Why is the null character not allowed? "foo\0bar" is a valid string in Rust.

Reply:
I don't think we should allow it in RON though, since that would make usage in e.g. C really difficult.